### PR TITLE
Add optimization for Head(SortValues) -> nlargest/nsmallest

### PR DESCRIPTION
**Pull request by mrocklin**
It feels nice.  Good follow-on work would be to support set_index.

### CODE REVIEW

File: dask_expr/_shuffle.py [hunk 1]
```diff
@@ -740,6 +740,29 @@ def _lower(self):
         return SetPartition(self.frame, self._other, self.drop, divisions)
 
     def _simplify_up(self, parent):
+        from dask_expr._expr import Head, Tail
+
+        # TODO, handle setting index with other frame
+        if (
+            isinstance(parent, Head)
+            and isinstance(self._other, (int, str))
```
Comment (phofl) [inline dask_expr/_shuffle.py:748]: Could we use the ``is_scalar`` check from dask_expr._util?

File: dask_expr/_shuffle.py [hunk 2]
```diff
@@ -740,6 +740,29 @@ def _lower(self):
         return SetPartition(self.frame, self._other, self.drop, divisions)
 
     def _simplify_up(self, parent):
+        from dask_expr._expr import Head, Tail
+
+        # TODO, handle setting index with other frame
```
Comment (phofl) [inline dask_expr/_shuffle.py:745]: Injecting an Assign is probably the easiest way of getting there

File: dask_expr/_shuffle.py [hunk 3]
```diff
@@ -740,6 +740,29 @@ def _lower(self):
...
+            and self._other in self.frame.columns
+        ):
+            return SetIndex(
+                NSmallest(self.frame, n=parent.n, _columns=self._other),
+                _other=self._other,
+            )
+
+        if (
+            isinstance(parent, Tail)
+            and isinstance(self._other, (int, str))
```
Comment (phofl) [inline dask_expr/_shuffle.py:758]: same here

File: dask_expr/_shuffle.py (hunk 1 above)
Comment (mrocklin) [reply dask_expr/_shuffle.py:748]:
```
(Pdb) pp self._other
df['x']
(Pdb) pp is_scalar(self._other)
True
```

Something is wrong here.   See also https://github.com/dask-contrib/dask-expr/issues/298

File: dask_expr/_shuffle.py (hunk 2 above)
Comment (mrocklin) [reply dask_expr/_shuffle.py:745]: Agreed.  I started doing this but it got complicated, so I stopped.

File: dask_expr/_shuffle.py (hunk 1 above)
Comment (phofl) [reply dask_expr/_shuffle.py:748]: I'll take a look, this worries me a bit

**Comment by mrocklin** [comment]
I've removed out the set_index stuff, which is more complicated.  I'm going to merge in the sort_values stuff and then raise another PR on top with set_index.  

This is mostly because I have a few other small things I want to raise and my branches on branches are getting weird.  

Sorry for the self merge.

**Pull request merged**